core/shared/src/main/scala/zio/stm/TReentrantLock.scala

```diff
@@ -0,0 +1,221 @@
+ *
+ * The two primary methods of this structure are `readLock`, which acquires a
+ * read lock in a managed context, and `writeLock`, which acquires a write lock
+ * in a managed context.
+ *
+ * Although located in the STM package, there is no need for locks within
+ * STM transactions. However, this lock can be quite useful in effectful code,
+ * to provide consistent read/write access to mutable state; and being in STM
+ * allows this structure to be composed into more complicated concurrent
+ * structures that are consumed from effectful code.
+ */
+class TReentrantLock private (data: TRef[Either[ReadLock, WriteLock]]) {
+
+  /**
+   * Acquires a read lock. The transaction will suspend until no other fibers
+   * are holding write locks. Succeeds with the number of write locks held by
+   * this fiber.
+   */
+  lazy val acquireRead: STM[Nothing, Int] =
+    STM.fiberId.flatMap(fiberId => adjustRead(fiberId, 1))
+
+  /**
+   * Acquires a write lock. The transaction will suspend until no other
+   * fibers are holding read or write locks. Succeeds with the number of
+   * write locks held by this fiber.
+   */
+  lazy val acquireWrite: STM[Nothing, Int] =
+    for {
+      fiberId <- STM.fiberId
+      w <- data.get.collect {
+            case Left(readLock) if (readLock.noOtherHolder(fiberId)) =>
+              WriteLock(1, readLock.readLocks(fiberId), fiberId)
+
+            case Right(WriteLock(n, m, `fiberId`)) =>
+              WriteLock(n + 1, m, fiberId)
+          }
+      _ <- data.set(Right(w))
+    } yield w.writeLocks
+
+  /**
+   * Just a convenience method for applications that only need reentrant locks,
+   * without needing a distinction between readers / writers.
+   *
+   * See [[writeLock]].
+   */
+  final lazy val lock: Managed[Nothing, Int] = writeLock
+
+  /**
+   * Determines if any fiber has a read or write lock.
+   */
+  final def locked: STM[Nothing, Boolean] =
+    (readLocked zipWith writeLocked)(_ || _)
+
+  /**
+   * Obtains a read lock in a managed context.
+   */
+  final lazy val readLock: Managed[Nothing, Int] =
+    Managed.make(acquireRead.commit)(_ => releaseRead.commit)
+
+  /**
+   * Retrieves the number of acquired read locks.
+   */
+  final def readLocks: STM[Nothing, Int] = data.get.map(_.fold(_.readLocks, _.readLocks))
+
+  /**
+   * Determines if any fiber has a read lock.
+   */
+  final def readLocked: STM[Nothing, Boolean] = readLocks.map(_ > 0)
+
+  /**
+   * Releases a read lock held by this fiber. Succeeds with the outstanding
+   * number of read locks held by this fiber.
+   */
+  lazy val releaseRead: STM[Nothing, Int] =
+    STM.fiberId.flatMap(fiberId => adjustRead(fiberId, -1))
+
+  /**
+   * Releases a write lock held by this fiber. Succeeds with the outstanding
+   * number of write locks held by this fiber.
+   */
+  lazy val releaseWrite: STM[Nothing, Int] =
+    STM.fiberId.flatMap(
+      fiberId =>
+        data.modify {
+          case Right(WriteLock(1, m, `fiberId`)) =>
```

Comment:
Should you test here that the two `fiberId` are equal?

Reply:
That’s what the bzckticks do. Instead of binding a new value to fiberId, it checks that the value is equal to the existing fiberId value.

Very strange pattern matching syntax 🤷🏼‍♂️

Reply:
🤯